docs/proposals/kubelet-cri-logging.md

```diff
@@ -0,0 +1,262 @@
+path to the log of a container in a pod.
+
+```
+/var/log/pods/<podUID>/<containerName>_<instance#>.log
+```
+
+In CRI, this is implemented by setting the pod-level log directory when
+creating the pod sandbox, and passing the relative container log path
+when creating a container.
+
+```
+PodSandboxConfig.LogDirectory: /var/log/pods/<podUID>/
+ContainerConfig.LogPath: <containerName>_<instance#>.log
+```
+
+Because kubelet determines where the logs are stores and can access them
+directly, this meets requirement (1). As for requirement (2), the log collector
+can easily extracts all necessary metadata (e.g., pod UID, container name) from
+the paths, and watch the directly for any changes.
+
+**Log format**
+
+The runtime should decorate each log entry with a RFC 3339 timestamp prefix,
+and ends with a newline.
+
+```
+2016-10-06T00:17:09.669794202Z The content of the log entry
+```
+
+With the knowledge, kubelet can parses the logs and serve them for `kubectl
+logs` requests. This meets requirement (3). Note that the format is defined
+deliberately simple to provide only information necessary to serve the requests.
+We do not intend for kubelet to host various logging plugins. It is also worth
+mentioning again that the scope of this proposal is restricted to stdout/stderr
+streams of the container, and we impose no restriction to the logging format of
+arbitrary container logs.
+
+**Who should rotate the logs?**
+
+We assume that a separate task (e.g., cron job) will be configured on the node
+to rotate the logs periodically, similar to today’s implementation.
+
+We do not rule out the possibility of letting kubelet or a per-node daemon
+(`DaemonSet`) to take up the responsibility, or even declare rotation policy
+in the kubernetes API as part of the `PodSpec`, but it is beyond the scope of
+the this proposal.
+
+**What about non-supported log formats?**
+
+If a runtime chooses to store logs in non-supported formats, it essentially
+opts out of `kubectl logs` features, which is backed by kubelet today. It is
+assumed that the user can rely on the advanced, cluster logging infrastructure
+to examine the logs.
+
+It is also possible that in the future, `kubectl logs` can contact the cluster
+logging infrastructure directly to serve logs [1a]. Note that this does not
+eliminate the need to store the logs on the node locally for reliability.
+
+
+**How can existing runtimes (docker/rkt) comply to the logging requirements?**
+
+In the short term, the ongoing docker-CRI integration [3] will support the
+proposed solution only partially by (1) creating symbolic links for kubelet
+to access, but not manage the logs, and (2) add support for json format in
+kubelet. A more sophisticated solution that either involves using a custom plugin
+or launching a separate process to copy and decorate the log will be considered
+as a mid-term solution.
+
+For rkt, there is a pending PR to enable systemd to redirect stdout/stderr to
+user-provided streams [4].
```

Comment:
A bit more context on this:
```
For rkt, implementation will rely on providing external file-descriptors for stdout/stderr
to applications via systemd [4]. Those streams are currently managed by a journald
sidecar, which collects stream outputs and store them in the journal file of the pod.
This will replaced by a custom sidecar which can produce logs in the format expected
by this specification and can handle clients attaching as well.
```